client/src/JobPaginatorState.js

```diff
@@ -0,0 +1,305 @@
+
+const countMagicLabels = (order, startAt) => {
+    let count = 0;
+    for (let startIndex = startAt; startIndex < order.length; startIndex++) {
+        if (!getLabel(order, startIndex)) {
+            count++;
+        }
+    }
+    return count;
+};
+
+const validateSave = (order) => {
+    const count = countMagicLabels(order, 0);
+    return count > 0 && !confirmSavedMagicLabels(count);
+};
+
+const useJobPaginator = (initialCategory, initialJob) => {
+    const [active, setActive] = useState(false);
+    const [currentPage, setCurrentPage] = useState(0);
+    const [zoom, setZoom] = useState(false);
+    const [order, setOrder] = useState([]);
+    const [magicLabelCache, setMagicLabelCache] = useState([]);
+    const [category, setCategory] = useState(initialCategory);
+    const [job, setJob] = useState(initialJob);
+
+    const saveMagicLabels = () => {
+        order.forEach((o, orderIndex) => {
+            if (!getLabel(order, orderIndex)) {
+                setLabel(orderIndex, getMagicLabel(orderIndex));
+            }
+        });
+    };
+
+    const setLabel = (imageNumber, text) => {
+        setMagicLabelCache([]); // clear label cache whenever there is a change
+        if (getLabel(order, imageNumber)) {
+            return;
+        }
+        order[imageNumber]["label"] = text !== null && text.length === 0 ? null : text;
+        setOrder(order);
+        dispatchEvent(new Event("Prep.editted"));
+    };
+
+    const getMagicLabel = (imageNumber) => {
+        const label = getLabel(order, imageNumber);
+
+        if (!label) {
+            return label;
+        }
+        if (typeof magicLabelCache[imageNumber] === "undefined") {
+            magicLabelCache[imageNumber] = MagicLabeler.getLabel(imageNumber, getMagicLabel);
+            setMagicLabelCache(magicLabelCache);
+        }
+        return magicLabelCache[imageNumber];
+    };
+
+    const setPage = (page) => {
+        if (page >= 0 && page < order.length) {
+            setCurrentPage(page);
+        }
+    };
+
+    const nextPage = () => {
+        setPage(currentPage + 1);
+    };
+
+    const prevPage = () => {
+        setPage(currentPage - 1);
+    };
+
+    const updatePagesByStatus = (status) => {
+        const {
+            file_problems: { deleted, added },
+        } = status;
+        const message = [];
+        if (deleted.length > 0) {
+            message.push(deleted.length + " file(s) have been removed = require(the job since the last edit.\n)");
+            setOrder(getNoneRemovedPages(order, deleted));
+        }
+
+        if (added.length > 0) {
+            message.push(added.length + " file(s) have been added to the job since the last edit.\n");
+            setOrder(getAddedPages(order, added));
+        }
+        return message.length ? alert(message.join()) : undefined;
+    };
+
+    const loadJob = async () => {
+        const { order } = await getJob(category, job);
+        setOrder(order);
+        setCategory(category);
```

Comment:
Is there any point to the setCategory and setJob calls here? It seems to me they're just setting the values to their current values, but maybe I'm missing a subtlety.

Reply:
had the same thought. this was in the old code. I wasn't sure why it was there, so I kept it there in case I broke something.

Reply:
I believe that in the old code, the assignment was necessary because "category" was coming in as a function parameter and getting assigned to the state. In the new code, you're initializing the state via the "initialCategory" property, so there's no longer a need to reassign the value later.

Reply:
Okay. I removed it